core/src/main/java/com/datastax/oss/driver/api/core/AllNodesFailedException.java

```diff
@@ -17,87 +17,150 @@
 
 import com.datastax.oss.driver.api.core.cql.ExecutionInfo;
 import com.datastax.oss.driver.api.core.metadata.Node;
-import com.datastax.oss.driver.shaded.guava.common.base.Joiner;
+import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableList;
 import com.datastax.oss.driver.shaded.guava.common.collect.ImmutableMap;
-import com.datastax.oss.driver.shaded.guava.common.collect.Iterables;
 import edu.umd.cs.findbugs.annotations.NonNull;
 import edu.umd.cs.findbugs.annotations.Nullable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.LinkedHashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.Map.Entry;
+import java.util.Set;
 
 /**
  * Thrown when a query failed on all the coordinators it was tried on. This exception may wrap
  * multiple errors, use {@link #getErrors()} to inspect the individual problem on each node.
  */
 public class AllNodesFailedException extends DriverException {
 
+  /** @deprecated Use {@link #fromErrors(List)} instead. */
   @NonNull
+  @Deprecated
   public static AllNodesFailedException fromErrors(@Nullable Map<Node, Throwable> errors) {
     if (errors == null || errors.isEmpty()) {
       return new NoNodeAvailableException();
     } else {
-      return new AllNodesFailedException(ImmutableMap.copyOf(errors));
+      return new AllNodesFailedException(groupByNode(errors.entrySet()));
     }
   }
 
   @NonNull
-  public static AllNodesFailedException fromErrors(
-      @Nullable List<Map.Entry<Node, Throwable>> errors) {
-    Map<Node, Throwable> map;
+  public static AllNodesFailedException fromErrors(@Nullable List<Entry<Node, Throwable>> errors) {
     if (errors == null || errors.isEmpty()) {
-      map = null;
+      return new NoNodeAvailableException();
     } else {
-      ImmutableMap.Builder<Node, Throwable> builder = ImmutableMap.builder();
-      for (Map.Entry<Node, Throwable> entry : errors) {
-        builder.put(entry);
-      }
-      map = builder.build();
+      return new AllNodesFailedException(groupByNode(errors));
     }
-    return fromErrors(map);
   }
 
-  private final Map<Node, Throwable> errors;
+  private final Map<Node, List<Throwable>> errors;
 
+  /** @deprecated Use {@link #AllNodesFailedException(String, ExecutionInfo, Iterable)} instead. */
+  @Deprecated
   protected AllNodesFailedException(
       @NonNull String message,
       @Nullable ExecutionInfo executionInfo,
       @NonNull Map<Node, Throwable> errors) {
+    this(message, executionInfo, groupByNode(errors.entrySet()));
+  }
+
+  protected AllNodesFailedException(
+      @NonNull String message,
+      @Nullable ExecutionInfo executionInfo,
+      @NonNull Iterable<Entry<Node, List<Throwable>>> errors) {
```

Comment:
Note that this can't be a `Map<Node, List<Throwable>>` because the erased type `Map` would clash with the existing constructor signature.